app/controllers/callbacks_controller.rb

```diff
@@ -2,80 +2,90 @@
+    connection_pairs = user_connections.map do |c|
+      { identifier: c['id'], platform: c['type'] }
+    end
 
-      break if steam_id_used || discord_id_used
+    # Generate a query to find any Identity with the
+    # specific pairs of identifiers and platforms
+    # TODO: Extract to model?
+    found_identities = connection_pairs.inject(Identity.none) do |memo, pair|
+      memo.or(Identity.where(pair))
+    end.pluck(:identifier, :user_id).to_h
 
-      new_user = User.create
-      Identity.create(user_id: new_user.id, platform: 'steam', identifier: steam_id)
-      Identity.create(user_id: new_user.id, platform: 'discord', identifier: discord_id)
+    ## TODO: DRY these two conditions
+    if connection_pairs.select { |c| c[:platform] == 'steam' }.empty?
+      error_codes << 'discord-missing-steam'
+    end
 
-      return redirect_to SUCCESS_URL
+    if connection_pairs.select { |c| c[:platform] == 'discord' }.empty?
+      identity_rows << Identity.new(platform: 'discord', identifier: discord_id)
     end
 
-    error_codes.push('discord-missing-steam') if error_codes.empty?
+    user_connections.each do |connection|
+      platform = connection['type']
+      identifier = connection['id']
+      is_verified = connection['verified']
 
-    error_codes = error_codes.join(',')
-    redirect_to "#{FAILURE_URL}?errors=#{error_codes}"
-  end
+      if is_verified == false
+        error_codes << 'steam-not-verified' if platform == 'steam'
+        next
+      end
 
-  def receive_cfc_bot_callback
-    steam_id = params['steam_id']
-    discord_id = params['discord_id']
+      user_id = found_identities[identifier]
 
-    error_codes = []
+      if user_id.nil?
+        identity_rows << Identity.new(platform: platform, identifier: identifier)
+      end
+    end
 
-    steam_id_used = Identity.find_by(platform: 'steam', identifier: steam_id)
-    discord_id_used = Identity.find_by(platform: 'discord', identifier: discord_id)
+    # We have to handle the situation where they may have multiple Users created
+    users = User.includes(:identities)
+                .where(id: found_identities.values)
+                .order('created_at DESC')
 
-    error_codes.push('used-steam-id') if steam_id_used
-    error_codes.push('used-discord-id') if discord_id_used
+    if users.empty?
+      User.create(identities: identity_rows)
+    else
+      user = users.first
 
-    render json: { status: 'failure', errors: error_codes } if error_codes.any?
+      identity_rows.map { |row| row.user_id = user.id }
+      Identity.import identity_rows
 
-    new_user = User.create
-    Identity.create(user_id: new_user.id, platform: 'steam', identifier: steam_id)
-    Identity.create(user_id: new_user.id, platform: 'discord', identifier: discord_id)
+      # Combine all other Users into the first (oldest) User
+      users.first.consume_users!(users.slice(1..)) if users.size > 1
+    end
+
+    if error_codes.any?
```

Comment:
after line ~72 you've already created the new user/identities even if there was a failure code which is probably intentional but may want to add a comment here explaining as such